File: src/main/java/dev/lain/claudejb/mcp/Toon.java
package dev.lain.claudejb.mcp;

public final class Toon {

    public static final class Options {
        public final int indentSize;
        public final char delimiter;
        public final boolean strict;

        public Options(int indentSize, char delimiter, boolean strict) {
            this.indentSize = indentSize;
            this.delimiter = delimiter;
            this.strict = strict;
        }

        public static final Options DEFAULT = new Options(2, ',', true);
    }

    private Toon() {
    }

    public static String encode(Object value) {
        return encode(value, Options.DEFAULT);
    }

    public static String encode(Object value, Options options) {
        return new ToonEncoder(options).encode(value);
    }

    public static Object decode(String text) {
        return decode(text, Options.DEFAULT);
    }

    public static Object decode(String text, Options options) {
        return new ToonDecoder(text, options).decode();
    }
}

File: src/main/kotlin/dev/lain/claudejb/model/mcp/toon/Toon.kt
package dev.lain.claudejb.model.mcp.toon

import kotlinx.serialization.json.JsonElement

enum class ToonDelimiter(val symbol: Char) {
    COMMA(','),
    TAB('\t'),
    PIPE('|'),
}

data class ToonOptions(
    val indentSize: Int = 2,
    val delimiter: ToonDelimiter = ToonDelimiter.COMMA,
    val strict: Boolean = true,
)

class ToonException(message: String) : RuntimeException(message)

internal fun toonError(message: String): Nothing = throw ToonException(message)

object Toon {

    fun encode(value: JsonElement, options: ToonOptions = ToonOptions()): String = ToonEncoder(options).encode(value)

    fun decode(text: String, options: ToonOptions = ToonOptions()): JsonElement = ToonDecoder(text, options).decode()
}
